data/src/main/java/com/linkedin/data/schema/PdlBuilder.java

```diff
@@ -275,7 +280,26 @@ else if (value instanceof DataList)
     {
       return jsonCodec.listToString((DataList) value);
     }
-    else if (value instanceof String)
+
+    return toPrimitiveValue(value);
+  }
+
+  protected String toOrderedBySchemaJson(Object value, DataSchema schema, JacksonDataTemplateCodec jsonCodec)
```

Comment:
Similar to my other suggestion, I suggest collapsing `toOrderedBySchemaJson` and `toJson` into one unified interface which accepts a `DataSchema` that may be null or non-null (e.g. `toJson(value, jsonCodec, schema)`. For all these methods, the javadoc should clearly state how the ordering is affected by null vs non-null arguments.

Reply:
This isn't possible at the moment because the codec that both each code path accepts is different, and there does not appear to be a common method I can call. I've renamed `toOrderedBySchemaJson` to `toJson` though and added documentation for the behavior.

It might be possible to refactor the underlying codecs so that the PdlBuilder can call the same method on both, which would make this possible. Let me know if you think it's worth pursuing.